### PR TITLE
table/tables: migrate test-infra to testify for `index_test.go`

### LINKED ISSUE #27588: migrate test-infra to testify for `table/tables/index_test.go`

### PR DESCRIPTION
**Pull request by feitian124**
### What problem does this PR solve?

Issue Number: closes #27588

resolve above issue in 3 parts, this is part 0:

- [x] migrate `index_test.go`
- [ ] migrate `partition_test.go`
- [ ] migrate `tables_test.go`

### Release note

```release-note
None
```

### CODE REVIEW

**Comment by feitian124** [comment]
The following test cases take too long to finish:
[2021-08-24T15:47:43.269Z] PASS: tidb_test.go:878: tidbTestSerialSuite.TestTLSAuto	7.539s

/run-check_dev_2

**Comment by feitian124** [comment]
not sure what is the best practise to setup and teardown the shared variables.
i try to use `setup` `teardown` style in commit  aa633538287e8969f4144d7660eeb7143c26fb9b but unit test fails so i reverted it.

**Review by tisonkun — CHANGES_REQUESTED** [review]
You can comment in https://github.com/pingcap/tidb/issues/26866 and ask me for creating subtasks. I'd prefer avoid using part 0,1,2 where number is hard to understand.

Sorry for my previous bad example.

See https://github.com/pingcap/tidb/issues/26863 as an example.

File: table/tables/index_test.go [hunk 1]
```diff
@@ -347,28 +339,28 @@ func (s *testIndexSuite) TestMultiColumnCommonHandle(c *C) {
...
 	for _, idx := range []table.Index{idxUnique, idxNonUnique} {
 		key, _, err := idx.GenIndexKey(sc, idxColVals, commonHandle, nil)
-		c.Assert(err, IsNil)
+		require.Nil(t, err)
 		_, err = idx.Create(mockCtx, txn, idxColVals, commonHandle, nil)
-		c.Assert(err, IsNil)
+		require.Nil(t, err)
 		val, err := txn.Get(context.Background(), key)
-		c.Assert(err, IsNil)
+		require.Nil(t, err)
```
Comment (tisonkun) [inline table/tables/index_test.go:363]: You may use `require.NoError` instead. Ditto others.

File: table/tables/index_test.go [hunk 2]
```diff
@@ -39,33 +39,25 @@ import (
...
-var _ = Suite(&testIndexSuite{})
-
 type testIndexSuite struct {
 	s   kv.Storage
 	dom *domain.Domain
 }
 
-func (s *testIndexSuite) SetUpSuite(c *C) {
-	testleak.BeforeTest()
+var s = testIndexSuite{}
```
Comment (tisonkun) [inline table/tables/index_test.go:50]: where is this suite used?

File: table/tables/index_test.go [hunk 3]
```diff
@@ -39,33 +39,25 @@ import (
...
+
+func TestSetUpSuite(t *testing.T) {
```
Comment (tisonkun) [inline table/tables/index_test.go:52]: setup should not be a test, you should batch setup logic and use when necessary

File: table/tables/index_test.go [hunk 4]
```diff
@@ -381,28 +373,34 @@ func (s *testIndexSuite) TestMultiColumnCommonHandle(c *C) {
...
 	stmt, err := parser.New().ParseOneStmt(sql, "", "")
-	c.Assert(err, IsNil)
+	require.Nil(t, err)
 	tblInfo, err := ddl.BuildTableInfoFromAST(stmt.(*ast.CreateTableStmt))
-	c.Assert(err, IsNil)
+	require.Nil(t, err)
 	return tblInfo
 }
+
+func TestTearDownSuite(t *testing.T) {
```
Comment (tisonkun) [inline table/tables/index_test.go:402]: teardown should not be a test, you should batch setup logic and use when necessary

**Review by tisonkun — CHANGES_REQUESTED** [review]
Comments inline. Even if you migrate tests by script, please do a human check and avoid push the reviewer to review a mechanical patch with corner case uncovered.

File: table/tables/index_test.go (hunk 3 above)
Comment (feitian124) [reply table/tables/index_test.go:52]: what is `batch setup logic ` mean?

Comment (feitian124) [reply table/tables/index_test.go:52]: ok, thanks.

File: table/tables/index_test.go [hunk 5]
```diff
@@ -17,55 +17,31 @@ package tables_test
...
-func (s *testIndexSuite) TearDownSuite(c *C) {
-	s.dom.Close()
-	err := s.s.Close()
-	c.Assert(err, IsNil)
-	testleak.AfterTest(c)()
-}
-
-func (s *testIndexSuite) TestIndex(c *C) {
+func TestIndex(t *testing.T) {
+	//t.Parallel()
```
[the remark this answers is not in the document]
Comment (tisonkun) [reply table/tables/index_test.go:44]: @feitian124 any updates?

[the remark this answers is not in the document]
Comment (feitian124) [reply table/tables/index_test.go:44]: did not find the root case yet, but noticed below 2 ColumnInfo with same name. 
column name should unique IMO.  kind of busy in last several days, will check tonight.
```go
		Columns: []*model.ColumnInfo{
			{ID: 1, Name: model.NewCIStr("c2"), State: model.StatePublic, Offset: 0, FieldType: *types.NewFieldType(mysql.TypeVarchar)},
			{ID: 2, Name: model.NewCIStr("c2"), State: model.StatePublic, Offset: 1, FieldType: *types.NewFieldType(mysql.TypeString)},
		},
```

[the remark this answers is not in the document]
Comment (feitian124) [reply table/tables/index_test.go:44]: i think i find the reason.
`testkit.CreateMockStore(t)` underlying invoke `client, pdClient, cluster, err := unistore.New(opts.path)`,
which used the default path as `conf.Engine.DBPath`. So race condition may happend if `t.Parallel` is used.

```go
	if path == "" {
		var err error
		if path, err = os.MkdirTemp("", "tidb-unistore-temp"); err != nil {
			return nil, nil, nil, err
		}
		persistent = false
	}
```

[the remark this answers is not in the document]
Comment (tisonkun) [reply table/tables/index_test.go:44]: I don't think so. MkdirTemp accepts two arguments as directory and pattern, not the exactly filename.

Other test cases already run in parallel with `testkit.CreateMockStore(t)`

[the remark this answers is not in the document]
Comment (tisonkun) [reply table/tables/index_test.go:44]: If you don't find the reason soon, I don't think it is a blocker. Just move the cases to a `_serial_test.go` and proceed.

[the remark this answers is not in the document]
Comment (feitian124) [reply table/tables/index_test.go]: you are right, seems not the reason.
indeed blocked a few days, i will  move the cases to a _serial_test.go and continue.

**Review by tisonkun — APPROVED** [review]

**Review by tiancaiamao — APPROVED** [review]

**Review by xhebox — APPROVED** [review]

File: table/tables/index_test.go [hunk 6]
```diff
@@ -283,53 +47,56 @@ func (s *testIndexSuite) TestSingleColumnCommonHandle(c *C) {
...
 		_, err = tablecodec.DecodeIndexKV(key, unTouchedVal, 1, tablecodec.HandleDefault,
 			tables.BuildRowcodecColInfoForIndexColumns(idx.Meta(), tblInfo))
-		c.Assert(err, IsNil)
+		require.NoError(t, err)
 	}
 }
 
-func (s *testIndexSuite) TestMultiColumnCommonHandle(c *C) {
+func TestMultiColumnCommonHandle(t *testing.T) {
+	t.Parallel()
```
Comment (xiongjiwei) [inline table/tables/index_test.go:96]: please see https://github.com/pingcap/tidb/pull/29125#discussion_r740047956

**Pull request merged**